File: 332_dfs_and_heap.cpp
class Solution {
private:
    void dfs(string& start, map<string, priority_queue<string, vector<string>, greater<string>>>& m, list<string>& res) {
        while (!m[start].empty()) {
            string arrival = m[start].top();
            m[start].pop();
            dfs(arrival, m, res);
        }
        res.push_front(start);
    }
public:
    vector<string> findItinerary(vector<pair<string, string>> tickets) {
        map<string, priority_queue<string, vector<string>, greater<string>>> m;
        for (auto p : tickets) m[p.first].push(p.second);
        list<string> res;
        string start = "JFK";
        dfs(start, m, res);
        return vector<string>(res.begin(), res.end());
    }
};

File: 332_dfs_and_heap.java
class Solution {
    private LinkedList<String> path = new LinkedList<>();
    private HashMap<String, PriorityQueue<String>> m = new HashMap<>();
    public List<String> findItinerary(String[][] tickets) {
        for (String[] pair : tickets) {
            m.putIfAbsent(pair[0], new PriorityQueue<>());
            m.get(pair[0]).add(pair[1]);
        }
        dfs("JFK");
        return path;
    }
    private void dfs(String departure) {
        PriorityQueue<String> q = m.get(departure);
        while (q != null && !q.isEmpty()) {
            dfs(q.poll());
        }
        path.addFirst(departure);
    }
}